Detectors/EMCAL/reconstruction/src/CaloRawFitter.cxx

```diff
@@ -0,0 +1,288 @@
+                                                                        mAmpCut(4),
+                                                                        mNsamplePed(3),
+                                                                        mIsZerosupressed(false),
+                                                                        mName(name),
+                                                                        mNameShort(nameshort),
+                                                                        mAlgo(FitAlgorithm::NONE),
+                                                                        mL1Phase(0),
+                                                                        mAmp(0)
+{
+  for (int i = 0; i < constants::ALTROMAXSAMPLES; i++) {
+    mReversed[i] = 0;
+  }
+}
+
+void CaloRawFitter::setTimeConstraint(const int min, const int max)
+{
+
+  if ((min > max) || min > constants::ALTROMAXSAMPLES || max > constants::ALTROMAXSAMPLES) {
+    LOG(WARN) << "Attempt to set Invalid time bin range (Min , Max) = (" << min << ", " << max << "), Ingored";
+  } else {
+    mMinTimeIndex = min;
+    mMaxTimeIndex = max;
+  }
+}
+
+unsigned short CaloRawFitter::maxAmp(const unsigned short* data, const int length) const
+{
+  unsigned short tmpmax = data[0];
+
+  for (int i = 0; i < length; i++) {
+    if (tmpmax < data[i]) {
+      tmpmax = data[i];
+    }
+  }
+  return tmpmax;
+}
+
+void CaloRawFitter::selectSubarray(const double* data, int length, short maxindex,
+                                   int* first, int* last, int cut) const
+{
+  int tmpfirst = maxindex;
+  int tmplast = maxindex;
+  double prevFirst = data[maxindex];
+  double prevLast = data[maxindex];
+  bool firstJump = false;
+  bool lastJump = false;
+
+  while ((tmpfirst >= 0) && (data[tmpfirst] >= cut) && (!firstJump)) {
+    // jump check:
+    if (tmpfirst != maxindex) { // neighbor to maxindex can share peak with maxindex
+      if (data[tmpfirst] >= prevFirst) {
+        firstJump = true;
+      }
+    }
+    prevFirst = data[tmpfirst];
+    tmpfirst--;
+  }
+
+  while ((tmplast < length) && (data[tmplast] >= cut) && (!lastJump)) {
+    // jump check:
+    if (tmplast != maxindex) { // neighbor to maxindex can share peak with maxindex
+      if (data[tmplast] >= prevLast) {
+        lastJump = true;
+      }
+    }
+    prevLast = data[tmplast];
+    tmplast++;
+  }
+
+  // we keep one pre- or post- sample if we can (as in online)
+  // check though if we ended up on a 'jump', or out of bounds: if so, back up
+  if (firstJump || tmpfirst < 0)
+    tmpfirst++;
+  if (lastJump || tmplast >= length)
+    tmplast--;
+
+  *first = tmpfirst;
+  *last = tmplast;
+
+  return;
+}
+
+float CaloRawFitter::reverseAndSubtractPed(const AltroDecoder::Bunch* bunch,
+                                           std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2,
+                                           double* outarray) const
```

Comment:
Raw pointer dangerous as ownership is not clear from the function definition (could be nullptr), better use std::vector<double> & the output array (and preferably as return value within a tuple) - the ownership would be clear from the reference.

Reply:
Parameter bunch is required in this function, it would be better to use a reference instead of a pointer (references cannot be nullptr). Applies to several other functions as well.